src/StoreApi/Utilities/OrderController.php

```diff
@@ -152,6 +152,40 @@ public function validate_order_before_payment( \WC_Order $order ) {
 				]
 			);
 		}
+
+		// Ensure all items on draft order are still in stock.
+		foreach ( $order->get_items() as $item ) {
+			$product = wc_get_product( $item->get_data()['product_id'] );
+
+			if ( ! $product->is_in_stock() ) {
+				throw new RouteException(
+					'woocommerce_rest_checkout_item_out_of_stock_error',
+					sprintf(
+					// Translators: %s Out of stock product.
+						__( 'Sorry, "%s" is not in stock. Please edit your basket and try again. We apologise for any inconvenience caused.', 'woo-gutenberg-products-block' ),
+						$product->get_name()
+					),
+					400
+				);
+			}
+
+			if ( $product->managing_stock() && ! $product->backorders_allowed() ) {
+				$qty_remaining = $product->get_stock_quantity();
+				$qty_on_order  = $item->get_quantity();
+
+				if ( $qty_remaining < $qty_on_order ) {
+					throw new RouteException(
+						'woocommerce_rest_checkout_item_out_of_stock_error',
+						sprintf(
+						// Translators: %s Not enough stock product.
+							__( 'Sorry, we do not have enough "%s" left in stock. Please edit your basket and try again. We apologise for any inconvenience caused.', 'woo-gutenberg-products-block' ),
+							$product->get_name()
+						),
+						400
+					);
+				}
+			}
+		}
```

Comment:
Code looks reasonable, though I wonder if the `foreach product { check stock }` logic could be refactored so it's shared. Were we doing this elsewhere already, or are we calling through to logic in core? 

Either way, just checking if it's worthwhile sharing this code, and the error messages.

Reply:
This makes sense, and I agree. I've just seen [something you added here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/StoreApi/Routes/Checkout.php#L362) and now I'm wondering if making changes to core, in the [`reserve_stock_for_order`](https://github.com/woocommerce/woocommerce/blob/master/src/Checkout/Helpers/ReserveStock.php#L66-L125) method so that these checks are done there and more appropriate error messages are returned (as discussed in pca54o-DN-p2) would be better than doing the checks in our plugin?

Reply:
Good question! That sounds like it's worth an investigation. There could be complexities caused by depending on this in Woo core (e.g. which version it's available vs what we support), but overall if we can work towards sharing this code and the error messages that's a good thing.